### DISCUSSION TITLE
Not connecting to FreeBSD

### DISCUSSION THREAD

**Discussion by dca00** [sshnet/SSH.NET#1035] [General]
Getting exception:

`"No suitable authentication method found to complete authentication (publickey,keyboard-interactive)."`

The same host through PuTTY, with the same credentials:

```
login as: root
Keyboard-interactive authentication prompts from server:
| Password for root@freebsd131:
End of keyboard-interactive prompts from server
Last login: Sun Nov  6 10:44:41 2022 from 192.168.1.32
FreeBSD 13.1-RELEASE releng/13.1-n250148-fc952ac2212 GENERIC

Welcome to FreeBSD!
```

**Comment by darkoperator** [comment]
You are opening issues twice for the same thing, don’t don that. You need to create an event handler and pass the correct connect info for keyboard interactive https://github.com/darkoperator/Posh-SSH/blob/7f6da8029338ffc972a1348a894fefc2c0c9d15b/Source/PoshSSH/PoshSSH/NewSessionBase.cs#L211 you can pass more than one connectioninfo type

**Comment by dca00** [comment]
Not sure I follow. Why does Connect() work for RHEL, Fedora, Debian, Ubuntu, and SUSE but not for FreeBSD?

**Comment by darkoperator** [comment]
Because those in sshd configuration file are PreferredAuthentications=passwordand and FreeBSD has it PreferredAuthentications=keyboard-interactive

**Comment by dca00** [comment]
It makes this library unworkable because we do not always know which OS, distro, or version we connect to. We have no way of knowing what are server settings. All of this has to be supported by the library, OOB because it may be **THE ONLY WAY** to connect to the box. If it depends on the box settings then this library has a very limited usability. Bottom line, this has to be the default behavior that it connects to anything and everything, like PuTTY does. PuTTY is your golden standard, and this library has to match it, if its developer has a hope of adoption.

**Comment by darkoperator** [comment]
You can specify multiple connection infos and let it choose which one to use. It takes an array

**Comment by darkoperator** [comment]
So you can specify password, multiple keyboard interactive for different prompts handled as events and even keys. As the server requests from one type to then other it changes to the correct one if you specified a connection info for the type. For my PS module I only use one but for version 4.x I will add multiple so it works when OTP is set or combo of key and password is used at some point

**Comment by dca00** [comment]
Interesting, but I do not understand anything that you write. I do not know what PS or OTP is, and what connection infos are. When one connects to SSH they know the IP, user name, and password or key. They do not need to know anything else, like in the case of PuTTY.

I glanced at your link. I have no faintest idea what your code does because SshClient does not expose any of methods/properties that you use. ConnectionInfo is an R/O property. What are you alluding to when you write that I have to create an event handler? For which event of SshClient? Are we referring to the same Nuget package SSH.NET or to different packages?

**Comment by darkoperator** [comment]
You are confusing an app with a library that you code against. In the library you need to manage all those parameters in your code. The link I sent you is C# code you could read and understand more how to use the library.

**Comment by dca00** [comment]
You've sent me something coded against an API that I do not find in SSH.NET.

**Comment by dca00** [comment]
Ah,I see! The event you had cursorily mentioned is not on SshClient. It is 
on an object
that  is passed
to its constructor.

This is not the most elegant solution because it increases the surface of SSH.NET at the consuming application.
Besides that your particular code would not have worked in 100% case because password prompts are not guaranteed to contain English "password" literal. They may be as short as ':' or '>' or whatever the vendor decides to use. I as a developer of an application that consumes SSH.NET have no way of knowing what kinds of passwords prompt the application may come across. This is why I contend that the login has to be handled in the library.

If we compare all of the above with WinScp that only needs a host address, a user name, and a password, then this is your golden standard if PuTTY is not good enough for you.
